File: python/python/ch3/TreeNode.py
class TreeNode:
    def __init__(self, val):
        self._val = val
        self._parent = None;
        self._children = []
        
    def is_root(self):
        return self._parent == None
      
    @property
    def children(self):
        return self._children
      
    def has_children(self):
        return len(self._children) > 0
      
    @property
    def value(self):
        return self._val
      
    @value.setter
    def value(self, value):
        self._val = value
        
    def append(self, child):
        child._parent = self
        self._children.append(child)
        return self
      
    def to_string(self):
        return f'Node (val: {self._val}, children: {len(self._children)})'

File: javascript/ch3/TreeNode.js
class TreeNode {
  constructor(val) {
    this._val = val;
    this._parent = null;
    this._children = [];
  }

  isRoot() {
    return this._parent == null;
  }

  get children() {
    return this._children;
  }

  hasChildren() {
    return this._children.length > 0;
  }

  get value() {
    return this._val;
  }

  set value(val) {
    this._val = val;
  }

  append(child) {
    child._parent = this;
    this._children.push(child);
    return this;
  }

  toString() {
    return `Node (val: ${this._val}, children: ${this._children.length})`;
  }
}

module.exports = {
  TreeNode
};